apps/labs/posts/unlocking-c-level-performance-in-df-apply.md

```diff
@@ -0,0 +1,463 @@
+
+```python
+%timeit df.sub(df.mean(axis=1), axis=0).div(df.std(ddof=0, axis=1), axis=0)
+```
+
+    2.51 ms ± 30.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
+
+In this case, parallel apply with raw=True provided a good speedup, however,
+this is not guaranteed to happen in all cases.
+
+A good rule of thumb to follow is that if there already exists a numpy/pandas function
+that does what you want to do already, you should use that, as it is hard for numba
+to beat the optimized low-level routines in both of those libraries.
+
+However, if your operation chains together a lot of these operations, numba may provide a pretty
+good speedup. For example, in our normalization example above, we took the mean, did a subtraction,
+took the standard deviation, and then did a division. Because numpy evaluates each one of these operations
+eagerly, it may miss out on performance optimizations that the numba compiler is able to use, explaining the
+performance improvement with numba above.
+
+```python
+# Let's take a look at the performance of just a mean operation with numba
+# (both raw mode and parallel on)
+# and just with numpy/pandas
+
+f = lambda x: x.mean()
+%timeit df.apply(f, engine='numba', raw=True, engine_kwargs={'parallel': True})
+%timeit df.mean(axis=1)
+%timeit df.values.mean(axis=1)
+```
+
+    178 µs ± 60.9 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
+    1.1 ms ± 99.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
+    32.9 µs ± 2.74 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
+
+Here, we can see that although numba beats the pandas mean function (it's a bit slower than the numpy version because it does some extra work),
+it loses to the numpy mean function, illustrating the point above.
+
+## Conclusion
+
+All in all, the numba engine will offer a faster way for pandas users to run their apply functions with minimal changes, for functions containing standard pandas operations (think indexing, arithmetic, and using numpy functions on the data) in pandas 2.2 and above.
+
+It does this by "mocking" the Series object passed in to your function. This is necessary because numba code cannot access the methods on a
+Series like Python code can - each numba equivalent must have either rewrite or wrap the logic behind the original method.
+
+It also opens up the door to parallel `DataFrame.apply`, when operating on the raw values of the DataFrame, which is super exciting as this has been a highly requested feature in pandas over the years.
+
+While many features/methods may be missing from the mocked Series object currently (such is the nature of mocking) and the numba engine in general, it's expected that these will slowly be added in as the numba engine is adopted, and from contributions from users like you!
+
+## Acknoledgements
+
+I'd like to thank [Matthew Roeschke](https://github.com/mroeschke) for reviewing my PRs, and NASA for funding the grant to make this possible.
```

Comment:
Is there somewhere else where we acknowledge the NASA grant?

It'd be good to be able to copy that over to make sure I'm not leaving anyone out that was involved in getting the grant.

Reply:
In talks perhaps, but no other blog post yet. I suggest to add the following sentence: _"This work was supported by a grant to Pandas, scikit-learn, SciPy and NumPy under the NASA ROSES 2020 program."_

Reply:
Updated.